modyn/supervisor/internal/triggers/ensembletrigger.py

```diff
@@ -0,0 +1,230 @@
+
+        # The current candidate index for the next trigger.
+        # If the aggregation function returns True, we trigger. If not,
+        # the processing head is still increased to the next triggering
+        # index of a subtrigger. Indexes are based on the new_data batch.
+        processing_head = -1
+
+        self._update_outdated_next_trigger_indexes(
+            processing_head,
+            subtrigger_generators,
+            next_subtrigger_index_cache,
+        )
+
+        while True:
+            next_trigger_idx = self._find_next_trigger_index(
+                processing_head,
+                new_data,
+                subtrigger_decision_cache,
+                next_subtrigger_index_cache,
+                log,
+            )
+            if next_trigger_idx is None:
+                # complete remaining inform calls regardless of the decision, any of the decision will lead to a
+                # non triggering aggregation result
+
+                break
+
+            processing_head = next_trigger_idx
+
+            # reset state for next detection pass;
+            self._reset_subtrigger_decision_cache(subtrigger_decision_cache)
+            self._update_outdated_next_trigger_indexes(
+                processing_head,
+                subtrigger_generators,
+                next_subtrigger_index_cache,
+            )
+
+            yield next_trigger_idx
+
+        self.last_inform_decisions = subtrigger_decision_cache
+
+    def inform_new_model(self, most_recent_model_id: int) -> None:
+        for trigger in self.subtriggers.values():
+            trigger.inform_new_model(most_recent_model_id)
+
+    # --------------------------------------------------- Internal --------------------------------------------------- #
+
+    def _create_subtriggers(self, config: EnsembleTriggerConfig) -> dict[str, Trigger]:
+        subtriggers: dict[str, Trigger] = {}
+        for trigger_id, trigger_config in config.subtriggers.items():
+            if isinstance(trigger_config, TimeTriggerConfig):
+                subtriggers[trigger_id] = TimeTrigger(trigger_config)
+            elif isinstance(trigger_config, DataAmountTriggerConfig):
+                subtriggers[trigger_id] = DataAmountTrigger(trigger_config)
+            elif isinstance(trigger_config, DataDriftTriggerConfig):
+                subtriggers[trigger_id] = DataDriftTrigger(trigger_config)
+            elif isinstance(trigger_config, EnsembleTriggerConfig):
+                raise NotImplementedError("Nested ensemble triggers are not supported.")
+            else:
+                raise ValueError(f"Unknown trigger config: {trigger_config}")
+        return subtriggers
+
+    def _reset_subtrigger_decision_cache(self, subtrigger_decision_cache: dict[str, bool]) -> None:
+        subtrigger_decision_cache.clear()
+        subtrigger_decision_cache.update({trigger_name: False for trigger_name in self.subtriggers})
+
+    def _update_outdated_next_trigger_indexes(
```

Comment:
maybe: `_update_next_subtrigger_index_cache` to make more clear what is being updated?